api/agent/resource.go

```diff
@@ -140,10 +150,85 @@ func (a *resourceTracker) GetResourceTokenWaiterCount() uint64 {
 	return waiters
 }
 
+func (a *resourceTracker) allocResourcesLocked(memory, cpuQuota uint64, isAsync bool) (uint64, uint64, uint64, uint64) {
+
+	var asyncMem, syncMem uint64
+	var asyncCPU, syncCPU uint64
+
+	if isAsync {
+		// async uses async pool only
+		asyncMem = memory
+		asyncCPU = cpuQuota
+	} else {
+		// if sync fits async + sync pool
+		syncMem = minUint64(a.ramSyncTotal-a.ramSyncUsed, memory)
+		syncCPU = minUint64(a.cpuSyncTotal-a.cpuSyncUsed, cpuQuota)
+
+		asyncMem = memory - syncMem
+		asyncCPU = cpuQuota - syncCPU
+	}
+
+	return asyncMem, syncMem, asyncCPU, syncCPU
+}
+
+func (a *resourceTracker) getResourceTokenNB(memory uint64, cpuQuota uint64, isAsync bool) ResourceToken {
+	if !a.IsResourcePossible(memory, cpuQuota, isAsync) {
+		return &resourceToken{decrement: func() {}, err: CapacityFull}
+	}
+	memory = memory * Mem1MB
+
+	a.cond.L.Lock()
+	defer a.cond.L.Unlock()
+
+	if !a.isResourceAvailableLocked(memory, cpuQuota, isAsync) {
+		return &resourceToken{decrement: func() {}, err: CapacityFull}
+	}
+
+	asyncMem, syncMem, asyncCPU, syncCPU := a.allocResourcesLocked(memory, cpuQuota, isAsync)
+
+	a.ramAsyncUsed += asyncMem
+	a.ramSyncUsed += syncMem
+	a.cpuAsyncUsed += asyncCPU
+	a.cpuSyncUsed += syncCPU
+
+	return &resourceToken{decrement: func() {
+		a.cond.L.Lock()
+		a.ramAsyncUsed -= asyncMem
+		a.ramSyncUsed -= syncMem
+		a.cpuAsyncUsed -= asyncCPU
+		a.cpuSyncUsed -= syncCPU
+		a.cond.L.Unlock()
+	}}
```

Comment:
we need to broadcast from this too, even if we get the resource in NB fashion, somebody could be waiting in blocking fashion by specifying the flag and we should wake them up since this is a delta (i know it's not set up like this externally, it's on or off, but it's a bug waiting to happen and not hard to close).

`allocResourcesLocked` imo should just return the resourceToken and handle the ramAsync/ramSync/cpuAsync/cpuSync bits too, they're in multiple places now but it's the same logic, and we need to have the broadcast here as well imo so it would be identical.